library/core/src/ptr/mod.rs

```diff
@@ -1744,54 +1749,67 @@ pub const unsafe fn write_unaligned<T>(dst: *mut T, src: T) {
     }
 }
 
-/// Performs a volatile read of the value from `src` without moving it. This
-/// leaves the memory in `src` unchanged.
-///
-/// Volatile operations are intended to act on I/O memory, and are guaranteed
-/// to not be elided or reordered by the compiler across other volatile
-/// operations.
-///
-/// # Notes
-///
-/// Rust does not currently have a rigorously and formally defined memory model,
-/// so the precise semantics of what "volatile" means here is subject to change
-/// over time. That being said, the semantics will almost always end up pretty
-/// similar to [C11's definition of volatile][c11].
-///
-/// The compiler shouldn't change the relative order or number of volatile
-/// memory operations. However, volatile memory operations on zero-sized types
-/// (e.g., if a zero-sized type is passed to `read_volatile`) are noops
-/// and may be ignored.
+/// Performs a volatile read of the value from `src` without moving it.
+///
+/// Rust does not currently have a rigorously and formally defined memory model, so the precise
+/// semantics of what "volatile" means here is subject to change over time. That being said, the
+/// semantics will almost always end up pretty similar to [C11's definition of volatile][c11].
+///
+/// Volatile operations are intended to act on I/O memory. As such, they are considered externally
+/// observable events (just like syscalls), and are guaranteed to not be elided or reordered by the
+/// compiler across other externally observable events. With this in mind, there are two cases of
+/// usage that need to be distinguished:
+///
+/// - When a volatile operation is used for memory inside an [allocation], it behaves exactly like
+///   [`read`], except for the additional guarantee that it won't be elided or reordered (see
+///   above). This implies that the operation will actually access memory and not e.g. be lowered to
+///   a register access or stack pop. Other than that, all the usual rules for memory accesses apply
```

Comment:
hmm, if the memory is literally on the stack though, wouldn't popping it off the stack be a valid implementation? I think something like "won't be lowered to reusing data from a previous read, such as a register that performed a previous load on that memory" might be clearer, since it describes the actual contract?

Reply:
I think it wouldn't be valid because a stack pop implies changing the stack pointer as well. Because we're talking about memory on the stack, that naturally falls under the Rust AM case. So if a `read_volatile` was allowed to lower into a pop, we'd have an extra side-effect caused by the read, which is disallowed for volatile on Rust memory.

In addition, I locally set up some preliminary testing for x86_64 and AVR, and it doesn't seem like `read_volatile` generates stack pops for either architecture, even for local variables on the stack. This is not conclusive, of course, but because of the above, it probably shouldn't anyway.

Reply:
The Rust AM has no concept of a Stack Pointer, so a priori there's no reason why the compiler couldn't use such a lowering.

But if that's undesirable, we should get LLVM to document that they don't do that, and then document the same ourselves.

Reply:
Lowering into a pop sounds strange to me, intuitively I think of volatile operations mapping to an explicit load/store of some form, with the memory address as a specific part of it in some way. But that may be biased toward the embedded use-case of peripheral I/O. I won't dwell on it, and change the line as requested.

Reply:
@LuigiPiucco Based on what you said, I'm slightly worried that you imagine that the following is not allowed to move the stack pointer, because that would be a "side effect":
```rust
let x: i32 = 0;
```

...and my understanding is it absolutely is?

The C contract of `volatile` doesn't say anything about stack operations either. I agree that the compiler will usually execute the enhanced contract you describe, but that seems mostly incidental. If a compiler decided to arrange locals on the stack so it could choose `pop`s more often for the final read from the stack, that seems fine?

Reply:
That nattering aside, revised text looks good.

Reply:
Don't worry, I don't think that. I just got my wires crossed yesterday, and it seemed like the pop would be invalid when using volatile, only.

Just to mention it, when I rebase I'll squash the commit with your suggestion (along with the many fixups) into a single documentation commit.